### PR TITLE
Handle inventory on death

### PR DESCRIPTION
**Pull request by darshan3**
**Project Card**:
[Handle respawn on death](https://github.com/orgs/Terasology/projects/17#card-21937677)

**How to test**:  
1. Host a Light and Shadow game
2. Join the game using one more client instance
3. Choose one team with each of the players
4. Give "player one" items and blocks through cheat "give item"
5. Collect opposition flag using 'player one'.
4. Kill "player one" by repeatedly attacking it through "player two".
5. Try to collect dropped items and flag using "player two".

**Expected outcome**:
The killed player should get respawned at its home base with full health restored. All the items held by them should get dropped at the same location except the magic staff. Trying to collect the dropped flag sends it directly to home base.

### CODE REVIEW

File: src/main/java/org/terasology/ligthandshadow/componentsystem/controllers/PlayerDeathSystem.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    @ReceiveEvent(priority = EventPriority.PRIORITY_HIGH)
+    public void beforeDestroy(BeforeDestroyEvent event, EntityRef player, CharacterComponent characterComponent, AliveCharacterComponent aliveCharacterComponent) {
+        if (player.hasComponent(PlayerCharacterComponent.class)) {
+            event.consume();
+            String team = player.getComponent(LASTeamComponent.class).team;
+
+            Prefab staffPrefab = assetManager.getAsset(LASUtils.MAGIC_STAFF_URI, Prefab.class).orElse(null);
+
+            Vector3f startPosition = new Vector3f(player.getComponent(LocationComponent.class).getLocalPosition());
+            Vector3f impulse = Vector3f.zero();
```
Comment (iaronaraujo) [inline src/main/java/org/terasology/ligthandshadow/componentsystem/controllers/PlayerDeathSystem.java]: Just out of curiosity, what would happen if the impulse wasn't 0?

Comment (darshan3) [reply src/main/java/org/terasology/ligthandshadow/componentsystem/controllers/PlayerDeathSystem.java]: Impulse 0 drops the items at the same place. If we increasing its value, items will be thrown in that direction.

File: src/main/java/org/terasology/ligthandshadow/componentsystem/controllers/PlayerDeathSystem.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+            for (int slotNumber = 0; slotNumber <= inventorySize; slotNumber++) {
+                EntityRef slot = inventoryManager.getItemInSlot(player, slotNumber);
+                Prefab currentPrefab = slot.getParentPrefab();
+                if (currentPrefab != null && !currentPrefab.equals(staffPrefab)) {
+                    int count = inventoryManager.getStackSize(slot);
+                    player.send(new DropItemRequest(slot, player, impulse, startPosition, count));
+                }
+            }
+
+            player.send(new DoHealEvent(100000, player));
```
Comment (llvieira) [inline src/main/java/org/terasology/ligthandshadow/componentsystem/controllers/PlayerDeathSystem.java]: Same here, it'd be better to use a constant variable.

File: src/main/java/org/terasology/ligthandshadow/componentsystem/controllers/PlayerDeathSystem.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    @In
+    private WorldProvider worldProvider;
+
+    @In
+    private BlockManager blockManager;
+
+    @ReceiveEvent(priority = EventPriority.PRIORITY_HIGH)
+    public void beforeDestroy(BeforeDestroyEvent event, EntityRef player, CharacterComponent characterComponent, AliveCharacterComponent aliveCharacterComponent) {
+        if (player.hasComponent(PlayerCharacterComponent.class)) {
+            event.consume();
```
Comment (nihal111) [inline src/main/java/org/terasology/ligthandshadow/componentsystem/controllers/PlayerDeathSystem.java]: Add a comment above explaining that this event handler consumes the event before the player entity is destroyed, hence avoiding a death screen and respawning the player at base.

File: src/main/java/org/terasology/ligthandshadow/componentsystem/controllers/PlayerDeathSystem.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+            String team = player.getComponent(LASTeamComponent.class).team;
+
+            Prefab staffPrefab = assetManager.getAsset(LASUtils.MAGIC_STAFF_URI, Prefab.class).orElse(null);
+
+            Vector3f startPosition = new Vector3f(player.getComponent(LocationComponent.class).getLocalPosition());
```
Comment (nihal111) [inline src/main/java/org/terasology/ligthandshadow/componentsystem/controllers/PlayerDeathSystem.java]: Why do we call this the `startPosition`? If anything, it should be the `endPosition` :laughing: or better, `deathPosition`.

Comment (skaldarnar) [reply src/main/java/org/terasology/ligthandshadow/componentsystem/controllers/PlayerDeathSystem.java]: `deathPosition` sounds about right …

File: src/main/java/org/terasology/ligthandshadow/componentsystem/controllers/PlayerDeathSystem.java (hunk 1 above)
Comment (skaldarnar) [reply src/main/java/org/terasology/ligthandshadow/componentsystem/controllers/PlayerDeathSystem.java]: which might be a nice effect later (to scatter them in a small radius around the player). Could even use the direction of killing impact … 🤔

File: src/main/java/org/terasology/ligthandshadow/componentsystem/controllers/PlayerDeathSystem.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+            Vector3f impulse = Vector3f.zero();
+            int inventorySize = inventoryManager.getNumSlots(player);
+            for (int slotNumber = 0; slotNumber <= inventorySize; slotNumber++) {
+                EntityRef slot = inventoryManager.getItemInSlot(player, slotNumber);
+                Prefab currentPrefab = slot.getParentPrefab();
+                if (currentPrefab != null && !currentPrefab.equals(staffPrefab)) {
+                    int count = inventoryManager.getStackSize(slot);
+                    player.send(new DropItemRequest(slot, player, impulse, startPosition, count));
+                }
+            }
```
Comment (skaldarnar) [inline src/main/java/org/terasology/ligthandshadow/componentsystem/controllers/PlayerDeathSystem.java]: Could you please introduce a new method to keep the inventory handling in a concise spot. The `beforeDestroy` method should read as close to the following as possible (i.e., a high-level description of the semantics):

```java
public void beforeDestroy(…) {
    event.consume();
    dropItemsFromInventory(player, …);
    healPlayer(player);
    respawnPlayerAtBase(player);
}
```

`healPlayer` and `respawnPlayerAtBase` are already single lines of code, but for inventory a new method will help.

**Pull request merged**